errors/errors.go

```diff
@@ -15,209 +15,310 @@ package errors
+func (e *Error) WithVars(a ...any) *Error {
+	eCopy := *e
+	eCopy.Message = fmt.Sprintf(e.Message, a...)
+
+	return &eCopy
 }
 
-func newResourceInfo(rid *ResourceInfo, err error) *errdetails.ResourceInfo {
-	owner := resourceInfoDefaultOwner
-	if rid.Owner != "" {
-		owner = rid.Owner
-	}
-	return &errdetails.ResourceInfo{
-		ResourceType: rid.Type,
-		ResourceName: rid.Name,
-		Owner:        owner,
-		Description:  err.Error(),
-	}
+func (e *Error) WithDetails(details ...proto.Message) *Error {
+	e.Details = append(e.Details, details...)
+
+	return e
 }
 
 // *** GRPC Methods ***
 
 // GRPCStatus returns the gRPC status.Status object.
 func (e *Error) GRPCStatus() *status.Status {
-	var stErr error
-	ste := status.New(e.grpcStatusCode, e.description)
-	if e.resourceInfo != nil {
-		ste, stErr = ste.WithDetails(newErrorInfo(e.reason, e.metadata), newResourceInfo(e.resourceInfo, e.err))
-	} else {
-		ste, stErr = ste.WithDetails(newErrorInfo(e.reason, e.metadata))
+	stat := status.New(e.GrpcCode, e.Message)
+
+	// convert details from proto.Msg -> protoiface.MsgV1
+	var convertedDetails []protoiface.MessageV1
+	for _, detail := range e.Details {
+		if v1, ok := detail.(protoiface.MessageV1); ok {
+			convertedDetails = append(convertedDetails, v1)
+		} else {
+			log.Debugf("Failed to convert error details: %s", detail)
+		}
 	}
-	if stErr != nil {
-		return status.New(codes.Internal, fmt.Sprintf("failed to create gRPC status message: %v", stErr))
+
+	if len(e.Details) > 0 {
+		var err error
+		stat, err = stat.WithDetails(convertedDetails...)
+		if err != nil {
+			log.Debugf("Failed to add error details: %s to status: %s", err, stat)
+		}
 	}
 
-	return ste
+	return stat
+
 }
 
 // *** HTTP Methods ***
 
-// ToHTTP transforms the supplied error into
-// a GRPC Status and then Marshals it to JSON.
-// It assumes if the supplied error is of type Error.
-// Otherwise, returns the original error.
-func (e *Error) ToHTTP() (int, []byte) {
-	resp, err := protojson.Marshal(e.GRPCStatus().Proto())
-	if err != nil {
-		errJSON, _ := json.Marshal(fmt.Sprintf("failed to encode proto to JSON: %v", err))
-		return http.StatusInternalServerError, errJSON
-	}
+// JSONErrorValue implements the errorResponseValue interface.
+func (e *Error) JSONErrorValue() []byte {
+	grpcStatus := e.GRPCStatus().Proto()
 
-	return e.httpCode, resp
-}
+	// Make httpCode human readable
 
-// HTTPCode returns the value of the HTTPCode property.
-func (e *Error) HTTPCode() int {
-	if e == nil {
-		return http.StatusOK
+	// If there is no http legacy code, use the http status text
+	// This will get overwritten later if there is an ErrorInfo code
+	httpStatus := e.Tag
+	if httpStatus == "" {
+		httpStatus = http.StatusText(e.HttpCode)
 	}
 
-	return e.httpCode
-}
+	// Change output to match prior output to not break users
+	errJson := map[string]interface{}{
```

Comment:
Could you use a strongly-typed struct here rather than a map?

Reply:
I didn't want to go the struct route and have all the diff possible types defined bc there are so many err_details, and since I didn't want to do that for the err_details to be consistent I went with this approach.

Reply:
Fine to use a map for details, but `errJSON` itself seems to have just a handful of keys: errorCode (string? I think), message (string), details (map[string]any)